### PR TITLE
Add support for selective deployment

### PR DESCRIPTION
**Pull request by enchobelezirev**
#### Description: 
This commit addes a new process parameters modulesForDeployment and resourcesForDeployment
which enable the selective deployment of modules and resources.
If specified, the parameters are intersected with the modules from the deployment descriptor
and based on the result the deployment is made. The same applies for the resources.
If the parameters are not specified, the deployment is working as before - deploy all modules + resources.
The client of the deploy-service backed is responsible for passing the parameters to the backend.

### CODE REVIEW

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v1/ApplicationsCloudModelBuilder.java [hunk 1]
```diff
@@ -118,10 +123,15 @@ protected ApplicationEnvironmentCloudModelBuilder createApplicationEnvironmentCl
                 SupportedParameters.DEPENDENCY_TYPE, DEPENDENCY_TYPE_HARD)
             .stream()
             .filter(module -> shouldDeployModule(module, mtaModulesInArchive, deployedModules))
+            .filter(module -> isModuleSpecifiedForDeployment(module))
             .map(this::getApplication)
             .collect(Collectors.toList());
     }
 
+    private boolean isModuleSpecifiedForDeployment(Module module) {
+        return CollectionUtils.isEmpty(modulesSpecifiedForDeployment) || modulesSpecifiedForDeployment.contains(module.getName());
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v1/ApplicationsCloudModelBuilder.java]: Suni: leave only modules...contais(module.getName()))

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v1/ServicesCloudModelBuilder.java [hunk 2]
```diff
@@ -32,29 +33,39 @@
     private CloudServiceNameMapper cloudServiceNameMapper;
     protected final PropertiesAccessor propertiesAccessor;
     protected final DeploymentDescriptor deploymentDescriptor;
+    protected final List<String> resourcesSpecifiedForDeployment;
     protected UserMessageLogger userMessageLogger;
 
-    public ServicesCloudModelBuilder(DeploymentDescriptor deploymentDescriptor, PropertiesAccessor propertiesAccessor,
-        CloudModelConfiguration configuration) {
+    public ServicesCloudModelBuilder(DeploymentDescriptor deploymentDescriptor, List<String> resourcesSpecifiedForDeployment,
```
Comment (enchobelezirev) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/cf/v1/ServicesCloudModelBuilder.java]: Make the Builder to work with only one module or resource...

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/RebuildApplicationDeployModelStep.java [hunk 3]
```diff
@@ -35,17 +33,16 @@ protected StepPhase executeStep(ExecutionWrapper execution) throws Exception {
...
-            getStepLogger().debug(Messages.MTA_ARCHIVE_MODULES, mtaArchiveModules);
-            Set<String> deployedModuleNames = CloudModelBuilderUtil.getDeployedModuleNames(deployedModules);
-            getStepLogger().debug(Messages.DEPLOYED_MODULES, deployedModuleNames);
-            Set<String> mtaModules = StepsUtil.getMtaModules(execution.getContext());
-            getStepLogger().debug(Messages.MTA_MODULES, mtaModules);
-            // TODO: Build a cloud model only for the current application:
-            List<CloudApplicationExtended> modifiedApps = getApplicationsCloudModelBuilder(execution.getContext()).build(mtaArchiveModules,
-                mtaModules, deployedModuleNames);
+            DeploymentDescriptor descriptor = StepsUtil.getDeploymentDescriptor(execution.getContext());
+            Module applicationModule = descriptor.getModules1()
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/RebuildApplicationDeployModelStep.java]: Extract this in a method.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/RebuildApplicationDeployModelStep.java [hunk 4]
```diff
@@ -35,17 +33,16 @@ protected StepPhase executeStep(ExecutionWrapper execution) throws Exception {
...
+                .stream()
+                .filter(module -> app.getModuleName()
+                    .equals(module.getName()))
+                .findFirst()
+                .orElseThrow(() -> new SLException(MessageFormat.format(
+                    "Application module name {0} is not matching any of the modules in the deployment descriptor", app.getModuleName())));
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/RebuildApplicationDeployModelStep.java]: Add a constant for this in cf.process.Messages.

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/util/CloudModelBuilderUtil.java [hunk 5]
```diff
@@ -17,6 +21,8 @@
 
 public class CloudModelBuilderUtil {
 
+    private final static Predicate<Module> DEFAULT_MODULE_FILTER_PREDICATE = (module) -> true;
```
[the remark this answers is not in the document]
Comment (enchobelezirev) [reply com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/util/CloudModelBuilderUtil.java]: Removed.

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/util/CloudModelBuilderUtil.java [hunk 6]
```diff
@@ -73,11 +79,20 @@ private static ResourceType getResourceType(Resource resource, PropertiesAccesso
         return ResourceType.get(type);
     }
 
-    public static Set<String> getModuleNames(DeploymentDescriptor deploymentDescriptor) {
-        Set<String> deployedModuleNames = new TreeSet<>();
-        for (Module mtaModule : deploymentDescriptor.getModules1()) {
-            deployedModuleNames.add(mtaModule.getName());
+    public static Set<String> getModuleNames(DeploymentDescriptor deploymentDescriptor, List<String> moduleNamesForDeployment) {
```
[the remark this answers is not in the document]
Comment (enchobelezirev) [reply com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/util/CloudModelBuilderUtil.java]: Will move it.

File: com.sap.cloud.lm.sl.cf.core/src/test/java/com/sap/cloud/lm/sl/cf/core/cf/v1/CloudModelBuilderTest.java [hunk 7]
```diff
@@ -423,6 +429,15 @@ public void setUp() throws Exception {
...
         servicesBuilder = getServicesCloudModelBuilder(deploymentDescriptor, configuration);
+
+        modulesCalculator = new ModulesCloudModelBuilderContentCalculator(mtaArchiveModules, deployedApps, mtaModules,
+            Collections.emptyList(), getProperitesAccessor(), null);
+
+        resourcesCalculator = new ResourcesCloudModelBuilderContentCalculator(Collections.emptyList(), getProperitesAccessor(), null);
+    }
+
+    protected PropertiesAccessor getProperitesAccessor() {
+        return new HandlerFactory(1).getPropertiesAccessor();
```
[the remark this answers is not in the document]
Comment (enchobelezirev) [reply com.sap.cloud.lm.sl.cf.core/src/test/java/com/sap/cloud/lm/sl/cf/core/cf/v1/CloudModelBuilderTest.java]: +1

File: com.sap.cloud.lm.sl.cf.core/src/test/java/com/sap/cloud/lm/sl/cf/core/cf/v3/CloudModelBuilderTest.java [hunk 8]
```diff
@@ -105,31 +97,4 @@ protected PlatformMerger getPlatformMerger(Platform platform, DescriptorHandler
         return new com.sap.cloud.lm.sl.mta.mergers.v2.PlatformMerger((com.sap.cloud.lm.sl.mta.model.v3.Platform) platform,
             (com.sap.cloud.lm.sl.mta.handlers.v3.DescriptorHandler) handler);
     }
-
-    @Test
-    public void testWarnMessage() {
```
[the remark this answers is not in the document]
Comment (enchobelezirev) [reply com.sap.cloud.lm.sl.cf.core/src/test/java/com/sap/cloud/lm/sl/cf/core/cf/v3/CloudModelBuilderTest.java]: Will return it.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BuildCloudDeployModelStep.java [hunk 9]
```diff
@@ -91,16 +116,59 @@ protected StepPhase executeStep(ExecutionWrapper execution) {
         }
     }
 
+    private List<Resource> calculateResourcesForDeployment(DeploymentDescriptor deploymentDescriptor,
```
[the remark this answers is not in the document]
Comment (enchobelezirev) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BuildCloudDeployModelStep.java]: For better reading?

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/RebuildApplicationDeployModelStep.java (hunk 4 above)
Comment (enchobelezirev) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/RebuildApplicationDeployModelStep.java]: +1

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BuildCloudDeployModelStep.java [hunk 10]
```diff
@@ -53,10 +64,17 @@ protected StepPhase executeStep(ExecutionWrapper execution) {
             StepsUtil.setServiceKeysToCreate(execution.getContext(), serviceKeys);
 
             // Build a list of applications for deployment and save them in the context:
+            CloudModelBuilderContentCalculator<Module> modulesCloudModelBuilderContentCalculator = getModulesContentCalculator(execution,
```
[the remark this answers is not in the document]
Comment (enchobelezirev) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BuildCloudDeployModelStep.java]: Done

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BuildCloudDeployModelStep.java [hunk 11]
```diff
@@ -67,7 +85,14 @@ protected StepPhase executeStep(ExecutionWrapper execution) {
             StepsUtil.setCustomDomains(execution.getContext(), customDomainsFromApps);
             getStepLogger().debug(Messages.CUSTOM_DOMAINS, customDomainsFromApps);
 
-            List<CloudServiceExtended> allServices = getServicesCloudModelBuilder(execution.getContext()).build();
+            CloudModelBuilderContentCalculator<Resource> resourcesCloudModelBuilderContentCalculator = getResourcesCloudModelBuilderContentCalculator(
```
[the remark this answers is not in the document]
Comment (enchobelezirev) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/BuildCloudDeployModelStep.java]: Done.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/RebuildApplicationDeployModelStep.java (hunk 3 above)
Comment (enchobelezirev) [reply com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/RebuildApplicationDeployModelStep.java]: Done.

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/util/CloudModelBuilderUtil.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (enchobelezirev) [reply com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/util/CloudModelBuilderUtil.java]: Done.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/ProcessDescriptorStep.java [hunk 12]
```diff
@@ -101,6 +112,31 @@ protected StepPhase executeStep(ExecutionWrapper execution) {
         }
     }
 
+    private Set<String> getModuleNames(DeploymentDescriptor deploymentDescriptor, List<String> moduleNamesForDeployment) {
+        if (moduleNamesForDeployment.isEmpty()) {
+            return deploymentDescriptor.getModules1()
+                .stream()
+                .map(module -> module.getName())
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/ProcessDescriptorStep.java]: Method reference.

File: com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/ProcessDescriptorStep.java [hunk 13]
```diff
@@ -101,6 +112,31 @@ protected StepPhase executeStep(ExecutionWrapper execution) {
...
+            .collect(Collectors.toSet());
+    }
+
+    private List<String> findInvalidModulesSpecifiedForDeployment(DeploymentDescriptor descriptor, List<String> modulesForDeployment) {
+        if (modulesForDeployment.isEmpty()) {
+            return Collections.emptyList();
+        }
+        Set<String> deploymentDescriptorModuleNames = descriptor.getModules1()
+            .stream()
+            .map(module -> module.getName())
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/ProcessDescriptorStep.java]: Method reference.

**Pull request merged**